From 4ad1177422ef1f940fa4a2416ce1e84836276fe0 Mon Sep 17 00:00:00 2001
From: Harsh Vardhan Dwivedi <hdwivedi@codeaurora.org>
Date: Mon, 14 Jul 2014 23:46:07 -0600
Subject: [PATCH] Add option to print call graph

Pass "-mllvm -print-call-graph" to clang to invoke.

This will generate dot files for every file compiled.

There will be an extra Node0x12345678 type node generated
by LLVM that can be removed by a post processing script.

Author: Harsh Vardhan Dwivedi <hdwivedi@codeaurora.org>
Signed-off-by: David Peixotto <dpeixott@codeaurora.org>
Signed-off-by: Harsh Vardhan Dwivedi <hdwivedi@codeaurora.org>
Signed-off-by: Mark Charlebois <charlebm@gmail.com>
[Upstream renamed createCallGraphPrinterPass() to
 createCallGraphDOTPrinterPass()]
Signed-off-by: Simon Dardis <simon.dardis@imgtec.com>
---
 lib/Transforms/IPO/PassManagerBuilder.cpp  |    9 +++++++++
 include/llvm/Analysis/DOTGraphTraitsPass.h |    8 +++++++-
 2 files changed, 16 insertions(+), 1 deletion(-)

--- llvm.orig/lib/Transforms/IPO/PassManagerBuilder.cpp
+++ llvm/lib/Transforms/IPO/PassManagerBuilder.cpp
@@ -18,6 +18,7 @@
 #include "llvm/Analysis/BasicAliasAnalysis.h"
 #include "llvm/Analysis/CFLAliasAnalysis.h"
 #include "llvm/Analysis/GlobalsModRef.h"
+#include "llvm/Analysis/CallPrinter.h"
 #include "llvm/Analysis/Passes.h"
 #include "llvm/Analysis/ScopedNoAliasAA.h"
 #include "llvm/Analysis/TargetLibraryInfo.h"
@@ -70,6 +71,10 @@
 RunFloat2Int("float-to-int", cl::Hidden, cl::init(true),
              cl::desc("Run the float2int (float demotion) pass"));
 
+static cl::opt<bool>
+PrintCallGraph("print-call-graph", cl::Hidden, cl::ZeroOrMore,
+			   cl::desc("Print the call graph"));
+
 static cl::opt<bool> RunLoadCombine("combine-loads", cl::init(false),
                                     cl::Hidden,
                                     cl::desc("Run the load combining pass"));
@@ -188,6 +193,9 @@
 
 void PassManagerBuilder::populateModulePassManager(
     legacy::PassManagerBase &MPM) {
+  if (PrintCallGraph)
+    MPM.add(createCallGraphDOTPrinterPass());
+
   // Allow forcing function attributes as a debugging and tuning aid.
   MPM.add(createForceFunctionAttrsLegacyPass());
 
--- llvm.orig/include/llvm/Analysis/DOTGraphTraitsPass.h
+++ llvm/include/llvm/Analysis/DOTGraphTraitsPass.h
@@ -16,6 +16,8 @@
 
 #include "llvm/Analysis/CFGPrinter.h"
 #include "llvm/Pass.h"
+#include "llvm/IR/Module.h"
+#include <sstream>
 #include "llvm/Support/FileSystem.h"
 
 namespace llvm {
@@ -158,7 +160,8 @@
 
   bool runOnModule(Module &M) override {
     GraphT Graph = AnalysisGraphTraitsT::getGraph(&getAnalysis<AnalysisT>());
-    std::string Filename = Name + ".dot";
+    std::string ModuleFileName = M.getModuleIdentifier();
+    std::string Filename = ModuleFileName + "_.dot";
     std::error_code EC;
 
     errs() << "Writing '" << Filename << "'...";
